# Predicting Average House Prices in Madrid Neighbourhoods Using Socioeconomic Indicators

## Overall Project Aim:
To develop and evaluate a machine learning model that uses 2025 socioeconomic indicators to predict average housing prices across Madrid neighbourhoods.

## 1. Data Inspection

### Introduction
Two datasets are used in this project:

- Panel of Socioeconomic Indicators of Districts and Neighbourhoods of Madrid (indicators.csv)
- Madrid Average Declared Housing Price (€ per m²) by District and Neighbourhood (prices.csv)

This notebook focuses on exploring the raw Madrid socioeconomic indicator dataset. The dataset contains multiple socioeconomic indicators across several years and includes data at both district and neighbourhood level.

### Import Libraries

In [1]:
# Import the Pandas library for data manipulation and analysis
import pandas as pd

### Load Dataset

In [2]:
# Loading Dataset 1: Panel of Sociodemographic Indicators of Districts and Neighborhoods of Madrid
indicators = pd.read_csv("../data/indicators.csv",sep=";")

### Initial Inspection

In [3]:
# Display the first ten rows
indicators.head(10)

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
0,1,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),523
1,2,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),258
2,3,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,134881
3,4,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,67829
4,5,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,67052
5,6,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Edad Media,NaN,NaN,Edad,Edad media de la población,"44,03"
6,7,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,0-14,NaN,Personas,Población de 0 a 14 años,10554
7,8,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,15-29,NaN,Personas,Población de 15 a 29 años,21800
8,9,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,30-44,NaN,Personas,Población de 30 a 44 años,42857
9,10,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,45-64,NaN,Personas,Población de 45 a 64 años,38025


In [4]:
# Display the last five rows
indicators.tail()

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
116903,116904,2025,Ciudad de Madrid,18.0,Villa de Vallecas,NaN,NaN,2024,01/01/2024,Estadística Ayuntamiento de Madrid,Salud,Discapacidad reconocida,Número de mujeres,Feminización Discapacidad,Mujeres,Porcentaje,Mujeres por cada 100 hombres (feminización de ...,"106,3"
116904,116905,2025,Ciudad de Madrid,19.0,Vicálvaro,NaN,NaN,2024,01/01/2024,Estadística Ayuntamiento de Madrid,Salud,Discapacidad reconocida,Número de mujeres,Feminización Discapacidad,Mujeres,Porcentaje,Mujeres por cada 100 hombres (feminización de ...,"101,31"
116905,116906,2025,Ciudad de Madrid,20.0,San Blas-Canillejas,NaN,NaN,2024,01/01/2024,Estadística Ayuntamiento de Madrid,Salud,Discapacidad reconocida,Número de mujeres,Feminización Discapacidad,Mujeres,Porcentaje,Mujeres por cada 100 hombres (feminización de ...,"97,54"
116906,116907,2025,Ciudad de Madrid,21.0,Barajas,NaN,NaN,2024,01/01/2024,Estadística Ayuntamiento de Madrid,Salud,Discapacidad reconocida,Número de mujeres,Feminización Discapacidad,Mujeres,Porcentaje,Mujeres por cada 100 hombres (feminización de ...,"100,64"
116907,116908,2025,Ciudad de Madrid,NaN,NaN,NaN,NaN,2024,01/01/2024,Estadística Ayuntamiento de Madrid,Salud,Discapacidad reconocida,Número de mujeres,Feminización Discapacidad,Mujeres,Porcentaje,Mujeres por cada 100 hombres (feminización de ...,"110,04"


In [5]:
# Display ten random rows
indicators.sample(10)

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
46705,46706,2022,Ciudad de Madrid,11.0,Carabanchel,112.0,Opañel,2021,01/01/2021,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,Nacionalidad extranjera,Personas,Personas con nacionalidad extranjera mujeres,4141
73782,73783,2023,Ciudad de Madrid,21.0,Barajas,NaN,NaN,2023,01/01/2023,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,65-79,NaN,Personas,Población de 65 a 79 años,6699
53004,53005,2022,Ciudad de Madrid,20.0,San Blas-Canillejas,206.0,Rejas,2021,01/01/2021,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de los hogares,Número de Hogares,hombre solo mayor de 65 años,NaN,Hogares,Hogares con un hombre solo mayor de 65 años,137
39055,39056,2022,Ciudad de Madrid,2.0,Arganzuela,NaN,NaN,2020,01/01/2020,Estadística Ayuntamiento de Madrid,Población del distrito,Dinámica demográfica,Esperanza de vida,Hombres,NaN,Edad,Esperanza de vida al nacer hombres,79
51426,51427,2022,Ciudad de Madrid,17.0,Villaverde,175.0,Ángeles,2021,01/01/2021,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,Nacionalidad extranjera,Personas,Personas con nacionalidad extranjera hombres,2885
31510,31511,2021,Ciudad de Madrid,16.0,Hortaleza,165.0,Apóstol Santiago,2020,01/01/2020,Estadística Ayuntamiento de Madrid,Vivienda,NaN,Inmuebles,Uso residencial,Superficie media,m2,Superficie media construida (m2) inmuebles de ...,108
93958,93959,2024,Ciudad de Madrid,8.0,Fuencarral-El Pardo,NaN,NaN,2023,28/05/2023,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Más Madrid-Verdes Equo,Municipio,Votos,"Elecciones municipales, votos a Más Madrid-Ver...",19095
99263,99264,2025,Ciudad de Madrid,4.0,Salamanca,NaN,NaN,2024,01/01/2024,Portal de datos abiertos del Ayuntamiento de M...,Servicios y equipamientos Municipales,Educación,Número de equipamientos,Escuelas infantiles,Privadas,Equipamientos,Escuelas infantiles privadas,11
62762,62763,2023,Ciudad de Madrid,7.0,Chamberí,76.0,Vallehermoso,2022,01/01/2022,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,Nacionalidad española,Personas,Personas con nacionalidad española hombres,8053
1239,1240,2020,Ciudad de Madrid,2.0,Arganzuela,24.0,Legazpi,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,65 y más,NaN,Personas,Población de 65 años y más,1702


In [6]:
# Check the number of rows and columns
indicators.shape

(116908, 18)

In [7]:
# Display all column names
indicators.columns

Index(['Orden', 'Periodo panel', 'ciudad', 'cod_distrito', 'distrito',
       'cod_barrio', 'barrio', 'año', 'fecha_indicador', 'fuente_indicador',
       'categoría_1', 'categoría_2', 'indicador_nivel1', 'indicador_nivel2',
       'indicador_nivel3', 'unidad_indicador', 'indicador_completo',
       'valor_indicador'],
      dtype='str')

#### Column Translation Key

The original Spanish column names have been translated into English. (Source: Google translate)

Output of column names (English):

Index(['Order', 'Panel period', 'City', 'District code', 'District',
       'Neighborhood code', 'Neighborhood', 'Year', 'Indicator date', 'Indicator source',
       'Category 1', 'Category 2', 'Indicator level 1', 'Indicator level 2',
       'Indicator level 3', 'Indicator unit', 'Full indicator name',
       'Indicator value'],
      dtype='object')

### Observations

The dataset contains multi-year data at both district and neighbourhood level. Each row represents an individual indicator and its value, organised by indicator type and subcategory. For example, separate rows may represent the number of male and female inhabitants within the same district and year.

As the model will focus on 2025 data at neighbourhood level, the data requires filtering by year and geographical level, followed by restructuring to support feature selection and modelling.

### Data Types

In [8]:
# Display the data type of each column
indicators.dtypes

Orden                   int64
Periodo panel           int64
ciudad                    str
cod_distrito          float64
distrito                  str
cod_barrio            float64
barrio                    str
año                     int64
fecha_indicador           str
fuente_indicador          str
categoría_1               str
categoría_2               str
indicador_nivel1          str
indicador_nivel2          str
indicador_nivel3          str
unidad_indicador          str
indicador_completo        str
valor_indicador           str
dtype: object

In [9]:
# Display dataset information
indicators.info()

<class 'pandas.DataFrame'>
RangeIndex: 116908 entries, 0 to 116907
Data columns (total 18 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Orden               116908 non-null  int64  
 1   Periodo panel       116908 non-null  int64  
 2   ciudad              116908 non-null  str    
 3   cod_distrito        115105 non-null  float64
 4   distrito            115105 non-null  str    
 5   cod_barrio          84570 non-null   float64
 6   barrio              84570 non-null   str    
 7   año                 116908 non-null  int64  
 8   fecha_indicador     116908 non-null  str    
 9   fuente_indicador    116908 non-null  str    
 10  categoría_1         116908 non-null  str    
 11  categoría_2         100705 non-null  str    
 12  indicador_nivel1    116908 non-null  str    
 13  indicador_nivel2    103504 non-null  str    
 14  indicador_nivel3    48898 non-null   str    
 15  unidad_indicador    116908 non-null  str    


### Observations
valor_indicador (indicator value) is a key column containing the numerical value for each socioeconomic indicator. It is currently stored as a string and will need to be converted to a numerical data type (e.g float) before modelling.

In [10]:
# Summary statistics of the numerical columns
indicators.describe()

,Orden,Periodo panel,cod_distrito,cod_barrio,año
count,116908.000000,116908.000000,115105.000000,84570.000000,116908.000000
mean,58454.500000,2022.539561,10.777282,110.687194,2021.616776
std,33748.576972,1.713728,5.911289,58.583824,1.946883
min,1.000000,2020.000000,1.000000,11.000000,2008.000000
25%,29227.750000,2021.000000,6.000000,62.000000,2020.000000
50%,58454.500000,2023.000000,11.000000,111.000000,2022.000000
75%,87681.250000,2024.000000,15.000000,158.000000,2023.000000
max,116908.000000,2025.000000,21.000000,215.000000,2025.000000


In [11]:
# Summary statistics for all columns
indicators.describe(include="all")

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
count,116908.000000,116908.000000,116908,115105.000000,115105,84570.000000,84570,116908.000000,116908,116908,116908,100705,116908,103504,48898,116908,116908,116908
unique,NaN,NaN,1,NaN,21,NaN,131,NaN,31,12,19,53,46,226,65,41,381,28852
top,NaN,NaN,Ciudad de Madrid,NaN,Ciudad Lineal,NaN,Villaverde Alto - Casco Histórico de Villaverde,NaN,01/01/2023,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de personas,Mujeres,Municipio,Personas,Superficie (ha.),1
freq,NaN,NaN,116908,NaN,7183,NaN,707,NaN,17246,89172,31665,23170,29881,9010,8711,44180,918,5245
mean,58454.500000,2022.539561,NaN,10.777282,NaN,110.687194,NaN,2021.616776,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,33748.576972,1.713728,NaN,5.911289,NaN,58.583824,NaN,1.946883,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,1.000000,2020.000000,NaN,1.000000,NaN,11.000000,NaN,2008.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,29227.750000,2021.000000,NaN,6.000000,NaN,62.000000,NaN,2020.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,58454.500000,2023.000000,NaN,11.000000,NaN,111.000000,NaN,2022.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,87681.250000,2024.000000,NaN,15.000000,NaN,158.000000,NaN,2023.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Missing Values and Duplicates

In [12]:
# Count missing values
indicators.isnull().sum()

Orden                     0
Periodo panel             0
ciudad                    0
cod_distrito           1803
distrito               1803
cod_barrio            32338
barrio                32338
año                       0
fecha_indicador           0
fuente_indicador          0
categoría_1               0
categoría_2           16203
indicador_nivel1          0
indicador_nivel2      13404
indicador_nivel3      68010
unidad_indicador          0
indicador_completo        0
valor_indicador           0
dtype: int64

In [13]:
# Count duplicate rows
indicators.duplicated().sum()

np.int64(0)

In [14]:
# Count number of unique values in each column
indicators.nunique()

Orden                 116908
Periodo panel              6
ciudad                     1
cod_distrito              21
distrito                  21
cod_barrio               131
barrio                   131
año                       18
fecha_indicador           31
fuente_indicador          12
categoría_1               19
categoría_2               53
indicador_nivel1          46
indicador_nivel2         226
indicador_nivel3          65
unidad_indicador          41
indicador_completo       381
valor_indicador        28852
dtype: int64

### Observations 

Initial summary statistics are of limited value at this stage, as the data has not yet been cleaned. Missing values indicate that not every indicator has a corresponding district, neighbourhood, subcategory or specific indicator value. No duplicate rows were identified. The dataset contains 21 districts and 131 neighbourhoods, consistent with the official counts for Madrid. These initial observations will inform the subsequent data cleaning and feature selection processes.